Jsoncpp Addition and README Update

mifelix opened the pull request:
Dockerfile:
- Add jsoncpp installation and building automatically when building Docker image.
- Changed the jsoncpp cmake to include the specific architecture to resolve "file in wrong format" error.
- Added a link and export for "missing files" to build jsoncpp successfully, until the master branch of jsoncpp is updated.
- Changed GCC_VERSION.

get_files.sh:
- Removed because it was empty.

README:
- Added instructions on building the docker engine and setting it up.

mifelix committed:
Dockerfile:
- Add jsoncpp installation and building automatically when building Docker image.
- Changed the jsoncpp cmake to include the specific architecture to resolve "file in wrong format" error.
- Added a link and export for "missing files" to build jsoncpp successfully, until the master branch of jsoncpp is updated.
- Changed GCC_VERSION.

get_files.sh:
- Removed because it was empty.

README:
- Added instructions on building the docker engine and setting it up.

Commit:
Dockerfile:
- Removed redundant commands such as 'cd build' and 'cmake ..' due to already being executed with line 124's 'cmake' command.
- Created additional symbolic links to libraries because jsoncpp build script is not able to find them.
*Note: Need to fix the missing libraries on the Jsoncpp repository and do a pull request to avoid further 'missing library' issues in the future.

Derrekito commented on Dockerfile:
```diff
@@ -115,4 +110,31 @@ RUN make CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc TARGE
+#update and upgrade
+RUN apt-get update -y && apt-get upgrade -y
+
+WORKDIR /shared/
```
add a comment that we change workdir to /shared so that when you drop into a docker shell running an instance of this image, you start off on the mount point "/shared"

Probably don't need the trailing "/"

mifelix replied:
I added the comment and also removed the trailing "/".

Derrekito commented on Dockerfile:
```diff
@@ -115,4 +110,31 @@ RUN make CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc TARGE
+RUN ln -s /usr/aarch64-linux-gnu/lib/libgcc_s.so.1 /lib/libgcc_s.so.1
+RUN ln -s /usr/aarch64-linux-gnu/lib/libc.so.6 /lib/libc.so.6
+RUN ln -s /usr/aarch64-linux-gnu/lib/libm.so.6 /lib/libm.so.6
+RUN export LD_LIBRARY_PATH=/usr/aarch64-linux-gnu/lib
```
Probably not necessary for this use case, but consider

`RUN export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/aarch64-linux-gnu/lib`

as it's good practice to avoid clobbering LD_LIBRARY_PATH.

mifelix replied:
Good catch, I made the change and tested it to ensure it worked as intended.

Derrekito commented on Dockerfile:
```diff
@@ -115,4 +110,31 @@ RUN make CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc TARGE
+COPY jsoncpp /tmp/jsoncpp
+
+# USER root
+RUN chown -R ${USER}:${USER} /tmp/jsoncpp
```
I think I did this with OpenBlas because I kept running into problems. It's not ideal. Try and see if json cpp compiles properly without chown command.

Derrekito replied:
It's not ideal because it screws up the permission in the repo making it a pain to pull to update the repo, making it easier to delete jsoncpp and cloning again if I needed to.

mifelix replied:
I removed it from the JsonCpp section and tested it. It does compile properly without it.

Derrekito commented on Dockerfile:
```diff
@@ -115,4 +110,31 @@ RUN make CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc TARGE
 
-WORKDIR /shared/
+# copy jsoncpp repo to tmp
+COPY jsoncpp /tmp/jsoncpp
```
```bash
# copy OpenBLAS repo to tmp
COPY OpenBLAS /tmp/OpenBLAS
#USER root
RUN chown -R ${USER}:${USER} /tmp/OpenBLAS
WORKDIR /tmp/OpenBLAS
```

Bruh, add some return spaces. Try to separate the clauses so the reader can groupings of things we're doing instead of it being a giant blob. consider adding a comment explaining that we're building and installing jsoncpp, followed by all the things we need to do to compile and install it.

mifelix replied:
I added some spaces to separate sections and for easier readability. I also added some comments explaining each section. Let me know if there are additional comments necessary.

Derrekito commented on build_image.sh:
since we're doing this pull request, can you add a return line before the function?

mifelix replied:
Added a return line before the function.

Derrekito commented on README.md:
```diff
@@ -1,2 +1,88 @@
-Docker development environment for the NVIDIA TK1
+# Docker-CUDA-AGX
+Docker development environment for the NVIDIA Jetson AGX
+- Compatible for development for the NVIDIA Jetson Orin NX
```
Docker development environment for the NVIDIA Jetson Orin Series (Nano, NX, and AGX).


Something like that. Try not to have hanging bullets if you can avoid it. They get lonely.

Derrekito replied:
This could really benefit from an introduction. Consider mentioning the target usecase for the docker image, the support libraries (like jsoncpp, openblas) that are custom compiled for this image. Also mention which CUDA version it's sporting.

mifelix replied:
Good point, I specified that the Docker development environment is for the NVIDIA Jetson Orin Series (Nano, NX, and AGX). 

I added an introduction, which mentions the target usecase for the docker image, the support libraries (JsonCpp and OpenBLAS), and the CUDA version utilized.

Derrekito commented on README.md:
```diff
@@ -1,2 +1,88 @@
+Docker development environment for the NVIDIA Jetson AGX
+- Compatible for development for the NVIDIA Jetson Orin NX
+
+Jsoncpp is included in the installation. Please refer to the following instructions for installing the Docker Engine and setting up the Docker CUDA AGX. 
```
Clarify which repos are cloned by running the [whatever script it is].

Try making the second sentence more helpful. Something like the following sections explain how to setup [Docker on Ubuntu] (<-- where that is a link to the section) and [building the docker image] (<-- the other section link).

mifelix replied:
I clarified which repos (OpenBLAS and JsonCpp) are cloned by running `build_image`. I also changed that second sentence to be more helpful, which points to specific sections in the README that it mentions.

Derrekito commented on README.md:
````diff
@@ -1,2 +1,88 @@
+git clone git@github.com:Derrekito/docker-cuda-agx.git
+cd docker-cuda-agx
+``` 
+Run the `build_image`:
````
What does the build image do?

mifelix replied:
I added a point explaining what `build_image` does.

Derrekito commented on README.md:
````diff
@@ -1,2 +1,88 @@
+./build_image.sh
+```
+
+Run the `create_container` and give a `path/to/projects` directory of your preference: 
````
what does create_container do?

mifelix replied:
I added a point explaining what `create_container` does.

Derrekito commented on README.md:
````diff
@@ -1,2 +1,88 @@
+./create_container.sh
+```
+
+Run `run_container.sh` to run your environment: 
````
what does run_container do? how is it helpful?

mifelix replied:
I added a point explaining what the `run_container` does and how it's helpful.

Derrekito commented on README.md:
```diff
@@ -1,2 +1,88 @@
+
+## Installing Docker Engine on Ubuntu
+
+For more information, visit the [dockerdocs](https://docs.docker.com/engine/install/ubuntu/).
```
If the following information is taken from instructions in this link, let's say so - sorta like a citation.

mifelix replied:
That is a good point. I rephrased it so it's not beating around the bush and just says that the instructions are summarized from and to visit the provided link for more information.

Derrekito commented on README.md:
````diff
@@ -1,2 +1,88 @@
+Ensure the Docker Engine installation is successful by running the `hello-world` image:
+
+```bash
+sudo docker run hello-world
````
sudo? I thought the instructions add the current user to the docker group? Doesn't the user need to log out and back in first and then can run without sudo?

mifelix replied:
The first section of instructions instructs the user to use `sudo` to test that the docker engine installed correctly and works properly. 

The second second outlines the post-installation steps which includes adding the current user to the docker group and log out & log back in, so the user does not have to utilize `sudo` anymore. This is tested at the end of the section which instructs the user to run the same "Hello World" line, but without `sudo`.

Derrekito reviewed (changes requested):
Very good! It's getting there 📦 

Let me know if you're up for adding the changes.

Commit:
build_image.sh:
- Indented before the function for readability and organization.

Dockerfile:
- Separated sections with comments for organization and easier readability.
- Commented out the `chown` command because it does work without it for JsonCpp.
- Added a comment for `WORKDIR`.
- Changed `LD_LIBRARY_PATH` so it is not clobbered.

README:
- Changed the first sentence to mention it supports NVIDIA's AGX, Nano, and NX.
- Added an introduction which mentions the target use cases, supported libraries, and which CUDA version it utilizes.
- Clarified which repos are cloned by running the `build_image.sh` script.
- Rewrote the second paragraph to flow easier and included links for easier readability and organization.
- Included descriptions for what `build_image`, `create_container`, and `run_container` does.
- Rephrased two sections to say the instructions are summarized and cited from their respective wesbites.

Derrekito commented on Dockerfile:
```diff
@@ -115,4 +117,33 @@ RUN make CC=aarch64-linux-gnu-gcc FC=aarch64-linux-gnu-gfortran HOSTCC=gcc TARGE
 #RUN make CC=aarch64-linux-gnu-gcc-11 FC=aarch64-linux-gnu-gfortran-11 HOSTCC=gcc-11 TARGET=CORTEXA73 USE_OPENMP=1 RANLIB=ranlib PREFIX=/usr/local install
```
Delete?

Derrekito commented on README.md:
```diff
@@ -1,2 +1,89 @@
+# Docker-CUDA-Orin-Series
+Docker development environment for the NVIDIA Jetson Orin Series (Nano, NX, and AGX).
+
+This Docker development environment is designed for the NVIDIA Jetson Orin Series (Nano, NX, and AGX). It is purposed for developmental purposes, supporting a variety of projects for these devices, and providing custom-compiled support libraries such as JsonCpp and OpenBLAS. The environment utilizes CUDA version 11.8. 
```
Erroneous comma before "and?" Does it utilize CUDA 11.8? I'm not sure what is meant by "supporting a variety of projects."

Derrekito commented on README.md:
```diff
@@ -1,2 +1,89 @@
-# docker-cuda-tk1
-Docker development environment for the NVIDIA TK1
+# Docker-CUDA-Orin-Series
+Docker development environment for the NVIDIA Jetson Orin Series (Nano, NX, and AGX).
```
You repeat this in the next paragraph. So probably cut this sentence.

Derrekito commented on README.md:
```diff
@@ -1,2 +1,89 @@
+
+## Installing Docker Engine on Ubuntu
+
+The following instructions are abstracted and summarized, please visit [dockerdocs](https://docs.docker.com/engine/install/ubuntu/) for more information.
```
Abstracted and summarized? Also, no love for active voice? 😔

Derrekito commented on README.md:
````diff
@@ -1,2 +1,89 @@
+    docker run hello-world
+    ```
+
+## Docker CUDA Setup
````
It's more of an Orin Series compiling environment

Derrekito commented on README.md:
````diff
@@ -1,2 +1,89 @@
+cd docker-cuda-agx
+``` 
+Run the `build_image` to build the image:
+- `build_image` downloads necessary files, clones the OpenBLAS and JsonCpp repositories, and builds the Docker image with the tag "cudatools:11.8".
````
11.8?

Derrekito commented on README.md:
````diff
@@ -1,2 +1,89 @@
+```
+
+Run `run_container.sh` to run your environment:
+- `run_container.sh` starts the Docker container, providing an isolated environment with all the necessary tools and libraries pre-installed.
````
Tools for what?

Derrekito commented on build_image.sh:
```diff
@@ -6,6 +6,7 @@ KEYRING="cuda-keyring_1.0-1_all.deb"
 CUDA_REPO_INSTALLER="cuda-repo-ubuntu2004-11-8-local_11.8.0-520.61.05-1_amd64.deb"
```
I'm so confused. Maybe I'm super tired. Okay so we ARE using CUDA 11.8 but gcc-9??? Ugh too tired to fix my previous comments. Sowwy
